### PR TITLE
[SHELL32][SDK] Implement SHTestTokenPrivilegeW

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

Implementing missing features...
JIRA issue: N/A

## Proposed changes

- Implement `SHTestTokenPrivilegeW` function.
- Modify `shell32.spec`.

## TODO

- [x] Do build.

### CODE REVIEW

File: dll/win32/shell32/utils.cpp [hunk 1]
```diff
@@ -103,6 +103,68 @@ SHInvokePrivilegedFunctionW(
...
+        goto Quit;
+
+    if (GetTokenInformation(hTokenChoice, TokenPrivileges, pTokenPriv, dwLength, &dwLength))
+    {
+        cPrivs = pTokenPriv->PrivilegeCount;
+        for (iPriv = 0; !ret && iPriv < cPrivs; ++iPriv)
+        {
+            pPrivLuid = &pTokenPriv->Privileges[iPriv].Luid;
+            ret = (Luid.LowPart == pPrivLuid->LowPart) &&
+                  (Luid.HighPart == pPrivLuid->HighPart);
```
Comment (HBelusca) [inline dll/win32/shell32/utils.cpp:155]: It may be possible to use the `RtlEqualLuid` function here.

File: dll/win32/shell32/utils.cpp [hunk 2]
```diff
@@ -103,6 +103,68 @@ SHInvokePrivilegedFunctionW(
...
+/*************************************************************************
+ *                SHTestTokenPrivilegeW (SHELL32.236)
+ *
+ * @see http://undoc.airesoft.co.uk/shell32.dll/SHTestTokenPrivilegeW.php
+ */
+EXTERN_C
+BOOL WINAPI
+SHTestTokenPrivilegeW(_In_opt_ HANDLE hToken, _In_z_ LPCWSTR lpName)
+{
+    UINT iPriv, cPrivs;
```
Comment (HBelusca) [inline dll/win32/shell32/utils.cpp:115]:
```suggestion
```

File: dll/win32/shell32/utils.cpp [hunk 3]
```diff
@@ -103,6 +103,68 @@ SHInvokePrivilegedFunctionW(
...
+    dwLength = 0;
+    GetTokenInformation(hTokenChoice, TokenPrivileges, NULL, 0, &dwLength);
+
+    pTokenPriv = (TOKEN_PRIVILEGES *)LocalAlloc(0, dwLength);
+    if (!pTokenPriv)
+        goto Quit;
+
+    if (GetTokenInformation(hTokenChoice, TokenPrivileges, pTokenPriv, dwLength, &dwLength))
+    {
+        cPrivs = pTokenPriv->PrivilegeCount;
```
Comment (HBelusca) [inline dll/win32/shell32/utils.cpp:150]:
```suggestion
        UINT iPriv, cPrivs;
        cPrivs = pTokenPriv->PrivilegeCount;
```

File: dll/win32/shell32/utils.cpp [hunk 4]
```diff
@@ -103,6 +103,68 @@ SHInvokePrivilegedFunctionW(
...
+            return FALSE;
+
+        hTokenChoice = hNewToken;
+    }
+
+    if (!LookupPrivilegeValueW(NULL, lpName, &Luid))
+        return FALSE;
+
+    dwLength = 0;
+    GetTokenInformation(hTokenChoice, TokenPrivileges, NULL, 0, &dwLength);
```
Comment (HBelusca) [inline dll/win32/shell32/utils.cpp:142]: Note that this call might fail for other reasons, and failure needs to be checked for.

File: dll/win32/shell32/utils.cpp [hunk 5]
```diff
@@ -103,6 +103,68 @@ SHInvokePrivilegedFunctionW(
...
+ * @see http://undoc.airesoft.co.uk/shell32.dll/SHTestTokenPrivilegeW.php
+ */
+EXTERN_C
+BOOL WINAPI
+SHTestTokenPrivilegeW(_In_opt_ HANDLE hToken, _In_z_ LPCWSTR lpName)
+{
+    UINT iPriv, cPrivs;
+    LUID Luid, *pPrivLuid;
+    DWORD dwLength;
+    TOKEN_PRIVILEGES *pTokenPriv;
```
Comment (HBelusca) [inline dll/win32/shell32/utils.cpp:118]:
```suggestion
    PTOKEN_PRIVILEGES pTokenPriv;
```

File: dll/win32/shell32/utils.cpp [hunk 6]
```diff
@@ -103,6 +103,68 @@ SHInvokePrivilegedFunctionW(
...
+        hTokenChoice = hNewToken;
+    }
+
+    if (!LookupPrivilegeValueW(NULL, lpName, &Luid))
+        return FALSE;
+
+    dwLength = 0;
+    GetTokenInformation(hTokenChoice, TokenPrivileges, NULL, 0, &dwLength);
+
+    pTokenPriv = (TOKEN_PRIVILEGES *)LocalAlloc(0, dwLength);
```
Comment (HBelusca) [inline dll/win32/shell32/utils.cpp:144]:
```suggestion
    pTokenPriv = (PTOKEN_PRIVILEGES)LocalAlloc(LPTR, dwLength);
```

File: dll/win32/shell32/utils.cpp [hunk 7]
```diff
@@ -103,6 +103,68 @@ SHInvokePrivilegedFunctionW(
...
+
+    if (!hToken)
+    {
+        if (!SHOpenEffectiveToken(&hNewToken))
+            goto Quit;
+
+        if (!hNewToken)
+            return FALSE;
+
+        hTokenChoice = hNewToken;
```
Comment (HBelusca) [inline dll/win32/shell32/utils.cpp:135]: You can economize one variable by removing the redundant `hTokenChoice` and just using `hToken` in this code block and below.
So here in particular,
```suggestion
        hToken = hNewToken;
```

File: dll/win32/shell32/utils.cpp (hunk 2 above)
Comment (katahiromz) [reply dll/win32/shell32/utils.cpp]: Fixed.

File: dll/win32/shell32/utils.cpp (hunk 3 above)
Comment (katahiromz) [reply dll/win32/shell32/utils.cpp:151]: Fixed.

File: dll/win32/shell32/utils.cpp (hunk 4 above)
Comment (katahiromz) [reply dll/win32/shell32/utils.cpp]: OK.

File: dll/win32/shell32/utils.cpp (hunk 7 above)
Comment (katahiromz) [reply dll/win32/shell32/utils.cpp]: Fixed.

**Pull request merged**